### PR TITLE
Initialize BlockTemplatesController for block themes only

### LINKED ISSUE #46463: Prevent 'BlockTemplatesController' to be initialized in classic themes which don't support templates nor template parts
`BlockTemplatesController` is a class used to handle block templates and templates parts.

While several of its methods check for the theme to support block templates/block templates parts before running, there are several other methods which don't have those checks, which is easy to cause bugs (see https://github.com/woocommerce/woocommerce/pull/46429 and https://github.com/woocommerce/woocommerce/issues/46461).

This issue is about try to only initialize `BlockTemplatesController` when the theme supports block templates or block template parts. So `BlockTemplatesController` wouldn't be initialized in classic themes at all.

One challenge is that the `check_should_use_blockified_product_grid_templates` method needs to be run also in classic themes, so it should be moved out of this class:

https://github.com/woocommerce/woocommerce/blob/de54195c3ab3a091c3c406984622b61a6f8fa413/plugins/woocommerce/src/Blocks/BlockTemplatesController.php#L217-L235

### PR DESCRIPTION
**Pull request by WunderBart**
## What?
Closes https://github.com/woocommerce/woocommerce/issues/46463
Supersedes https://github.com/woocommerce/woocommerce/pull/48385

- Move the `wc_blocks_use_blockified_product_grid_block_as_template` option check from BTC to `wc-template-functions.php` as it's shared between classic and block themes.
- Effectively switch the entire BTC class to be only initialized for block themes. 

## How to test

For the `wc_blocks_use_blockified_product_grid_block_as_template` option part, ensure:
- it's switched on when switching from classic to block theme,
- stays switched on when switching from block to block theme,
- is switched off when switching to a classic theme,
- stays switched off when switching from classic to classic theme.

For the BTC initialization switch part:

- Generally, if the tests pass, it should be good to go. The only risk here is that we missed some functional leaks into classic themes, so we might want to do some manual smoke testing. Therefore, switch a classic theme on (e.g. Storefront) and ensure you can edit and view the content as expected.

### CODE REVIEW

**Comment by WunderBart** [comment]
Asked about the failing blocks E2Es here: https://github.com/woocommerce/woocommerce/pull/48224#issuecomment-2162987518

File: plugins/woocommerce/src/Blocks/Domain/Bootstrap.php [hunk 1]
```diff
@@ -153,12 +153,15 @@ function () {
...
 			$this->container->get( BlockTypesController::class );
 			$this->container->get( BlockTemplatesRegistry::class )->init();
-			$this->container->get( BlockTemplatesController::class )->init();
 			$this->container->get( ClassicTemplatesCompatibility::class );
 			$this->container->get( ArchiveProductTemplatesCompatibility::class )->init();
 			$this->container->get( SingleProductTemplateCompatibility::class )->init();
 			$this->container->get( Notices::class )->init();
 			$this->container->get( PTKPatternsStore::class );
+
+			if ( wc_current_theme_is_fse_theme() ) {
```
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/Domain/Bootstrap.php]: A classic theme might support block template parts, if I'm not wrong. So we should account for that:

```suggestion
			if ( wc_current_theme_is_fse_theme() || current_theme_supports( 'block-template-parts' ) ) {
```

File: plugins/woocommerce/src/Blocks/Domain/Bootstrap.php [hunk 2]
```diff
@@ -427,6 +424,15 @@ function () {
 				return new QueryFilters();
 			}
 		);
+
+		if ( wc_current_theme_is_fse_theme() ) {
```
Comment (Aljullu) [inline plugins/woocommerce/src/Blocks/Domain/Bootstrap.php]: Same here. :slightly_smiling_face: 

```suggestion
		if ( wc_current_theme_is_fse_theme() || current_theme_supports( 'block-template-parts' ) ) {
```

**Review by Aljullu — APPROVED** [review]
It's working great and code changes look good, thanks @WunderBart! Just a few questions:

1. Did you evaluate placing the `wc_check_should_use_blockified_product_grid_templates` function somewhere else? I don't have a better suggestion, but thinking that it would be really nice if we could keep it inside `/src/Blocks/`, given that it's related to block themes. Not sure how, thought, maybe in a new file that is only initialized in the editor? :thinking:  (This is not a hard opinion, just thinking out loud to know your thoughts)
2. I think the testing steps should be separated on what should be tested by GlobalStep and what shouldn't (probably the database part)? Btw, I think the option name is `wc_blocks_use_blockified_product_grid_block_as_template`

File: plugins/woocommerce/includes/wc-template-functions.php [hunk 3]
```diff
@@ -4017,4 +4018,25 @@ function wc_update_product_archive_title( $post_type_name, $post_type ) {
...
 
+/**
+ * Checks the old and current themes and determines if the "wc_blocks_use_blockified_product_grid_block_as_template"
+ * option need to be updated accordingly.
+ *
+ * @param string    $old_name Old theme name.
+ * @param \WP_Theme $old_theme Instance of the old theme.
+ * @return void
+ */
+function wc_check_should_use_blockified_product_grid_templates( $old_name, $old_theme ) {
```
Comment (tjcafferkey) [inline plugins/woocommerce/includes/wc-template-functions.php]: Should we not read the option first? Then we only need to update it if its different and is less expensive than writing so a slight optimization.

**Comment by WunderBart** [comment]
> Did you evaluate placing the wc_check_should_use_blockified_product_grid_templates function somewhere else? I don't have a better suggestion, but thinking that it would be really nice if we could keep it inside /src/Blocks/, given that it's related to block themes. 

Good question, @Aljullu. I figured it should be outside of the `src/Blocks` because, IIUC, it’s relevant to both block and classic themes in the same way: The check disables that option when we switch from a block theme to a classic theme and enables it when we switch from a block theme to a classic theme. Now I kinda realize that it didn't have to be moved all the way out of the `src/Blocks` though, only the BTC class. 🤔 

> Not sure how, thought, maybe in a new file that is only initialized in the editor? 🤔 (This is not a hard opinion, just thinking out loud to know your thoughts)

What if we moved that check to the [Options class](https://github.com/woocommerce/woocommerce/blob/2821bc2a8c363a26c45fb3c149b551cd66ee3adb/plugins/woocommerce/src/Blocks/Options.php) itself? Depending on the perspective, this could either support or contradict the separation of concerns. If we'd like to separate it, maybe some sort of `OptionsManager` class could handle the options setup. I'm unsure if we wouldn't be over-optimizing at this point, as just one option needs to be set up this way. Anyway, if the `OptionsManager` idea sounds good to you, it looks to me like we could initialize it from Bootstrap.php [somewhere along these lines](https://github.com/woocommerce/woocommerce/blob/4714137b2019bbbd60cc544992d860fc8321d3e8/plugins/woocommerce/src/Blocks/Domain/Bootstrap.php#L129-L136)?

File: plugins/woocommerce/includes/wc-template-functions.php (hunk 3 above)
Comment (WunderBart) [reply plugins/woocommerce/includes/wc-template-functions.php]: Sounds good! I like making things less expensive. 😄

File: plugins/woocommerce/src/Blocks/Domain/Bootstrap.php (hunk 1 above)
Comment (WunderBart) [reply plugins/woocommerce/src/Blocks/Domain/Bootstrap.php]: Great catch, and I'm kinda surprised that all the tests have passed without it. Should we add a basic test to cover this part?

Comment (Aljullu) [reply plugins/woocommerce/src/Blocks/Domain/Bootstrap.php]: Yup, good point! We definitely should. Do you want to open an issue and add it to the Project board?

Comment (WunderBart) [reply plugins/woocommerce/src/Blocks/Domain/Bootstrap.php]: Sounds good. OK if we add those tests along with this PR?

The issue is here: https://github.com/woocommerce/woocommerce/issues/48468

Comment (Aljullu) [reply plugins/woocommerce/src/Blocks/Domain/Bootstrap.php]: Sure thing!

**Comment by Aljullu** [comment]
> What if we moved that check to the [Options class](https://github.com/woocommerce/woocommerce/blob/2821bc2a8c363a26c45fb3c149b551cd66ee3adb/plugins/woocommerce/src/Blocks/Options.php) itself? Depending on the perspective, this could either support or contradict the separation of concerns. If we'd like to separate it, maybe some sort of `OptionsManager` class could handle the options setup. I'm unsure if we wouldn't be over-optimizing at this point, as just one option needs to be set up this way. Anyway, if the `OptionsManager` idea sounds good to you, it looks to me like we could initialize it from Bootstrap.php [somewhere along these lines](https://github.com/woocommerce/woocommerce/blob/4714137b2019bbbd60cc544992d860fc8321d3e8/plugins/woocommerce/src/Blocks/Domain/Bootstrap.php#L129-L136)?

That sounds like a good plan. It's not a big deal, but what I would really like is to:

1. have it inside `/src/Blocks` because even though it also applies to classic themes, the option itself is only relevant for block templates
2. ideally, only run the code in the admin site, not in the frontend (I know it's not a big deal because we use the `after_switch_theme` action, but still I think it reduces the chances of introducing bugs in the future)

So creating an `OptionsManager` class works for me. :100:

File: plugins/woocommerce/includes/wc-template-functions.php (hunk 3 above)
Comment (WunderBart) [reply plugins/woocommerce/includes/wc-template-functions.php]: Updated in https://github.com/woocommerce/woocommerce/pull/48382/commits/d9536dc4f8cc20dfe832e4b356a7372e47f1cc19

**Comment by WunderBart** [comment]
This PR unveiled an issue where the `current_theme_supports( 'block-template-parts' )` always returns `false` when called from the [`Bootstrap.php`](https://github.com/woocommerce/woocommerce/blob/6431a9c7d31bf434c1aa85f13741a404e47bc09b/plugins/woocommerce/src/Blocks/Domain/Bootstrap.php) class. I spent some time debugging this, and it seems that the Bootstrap class is initialized **before** the `after_setup_theme` hook is fired.

For testing purposes, I created a classic theme with the block template parts support and [uploaded it to this PR](https://github.com/woocommerce/woocommerce/blob/4eb8efecc4d8bf4dd49055f1544337f718ab32ad/plugins/woocommerce-blocks/tests/e2e/themes/storefront-child__with-block-template-parts). I've also pushed some [debugging logs for easier inspection](https://github.com/woocommerce/woocommerce/pull/48382/commits/4c54eba3690ce6312bc70144bc947686b8954860).

To verify the issue:

0. `cd plugins/woocommerce-blocks`.
1. Start wp-env with `npx wp-env start --udpate`.
2. Go to http://localhost:8889/wp-admin/themes.php.
3. Activate the `Storefront Child (with block template parts)` theme.
4. Check the debug.log, it should contain something like this:

   ```log
   [21-Jun-2024 14:48:06 UTC] -------------------------------
   [21-Jun-2024 14:50:07 UTC] From Bootstrap.php:
   [21-Jun-2024 14:50:07 UTC] Theme: Storefront Child (with block template parts)
   [21-Jun-2024 14:50:07 UTC] Supports BTP: no
   [21-Jun-2024 14:50:07 UTC] -------------------------------
   [21-Jun-2024 14:50:07 UTC] From theme's functions.php:
   [21-Jun-2024 14:50:07 UTC] Theme: Storefront Child (with block template parts)
   [21-Jun-2024 14:50:07 UTC] Supports BTP: yes
   [21-Jun-2024 14:50:07 UTC] -------------------------------
   ```

> [!NOTE]
> If the `wp-env` debug logs aren't showing, you might need to do the following:
> ```sh
> npx wp-env run tests-cli wp config set WP_DEBUG 1
> npx wp-env run tests-cli wp config set WP_DEBUG_LOG 1
> npx wp-env run tests-wordpress -- touch wp-content/debug.log
> npx wp-env run tests-wordpress -- chmod 666 wp-content/debug.log
> npx wp-env run tests-wordpress -- tail -f wp-content/debug.log
> ```

@Aljullu @gigitux @tjcafferkey, how do you suggest approaching this issue? Should we move the BTC initialization somewhere else? Or maybe the entire `Bootstrap.php`? 😅

**Comment by tjcafferkey** [comment]
Hmm, I am trying to think if there is another method in which we can reliably determine if we should initialize the BTC.

>Should we move the BTC initialization somewhere else? Or maybe the entire Bootstrap.php? 😅

I think if we can't find another way then it would make more sense/be safer to move the BTC initialization rather than the Bootstrap one. But only if we can't find another way around this limitation.

**Comment by tjcafferkey** [comment]
Would there be any danger to doing something similar [to this](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Blocks/Domain/Bootstrap.php#L114-L123) and initializing the BTC when the `after_setup_theme` hook is fired?

**Comment by Aljullu** [comment]
> Would there be any danger to doing something similar [to this](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Blocks/Domain/Bootstrap.php#L114-L123) and initializing the BTC when the `after_setup_theme` hook is fired?

Good idea! I'm trying this approach in https://github.com/woocommerce/woocommerce/pull/48905 and it seems to work fine, I did even test it with the modified Storefront version @WunderBart created and I couldn't spot anything off. We can either hook into `after_setup_theme` with a high priority or into `init` with any priority, I lean towards the second approach, but I don't have a strong opinion.

I will continue working on this tomorrow, as there are some tests failing, but hopefully I'll be able to open the PR for review before EOW. :crossed_fingers:

**Comment by Aljullu** [comment]
> We can either hook into `after_setup_theme` with a high priority or into `init` with any priority, I lean towards the second approach, but I don't have a strong opinion.

Well, not really. We need to use `after_setup_theme` as `init` seems to be too late and causes some issues with template parts.

In any case, I opened the PR for review in https://github.com/woocommerce/woocommerce/pull/48905. :slightly_smiling_face:

**Comment by WunderBart** [comment]
Closing in favor of https://github.com/woocommerce/woocommerce/pull/48905. Thanks @Aljullu and @tjcafferkey! 🙇

**Pull request closed without merging**